Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/test/core/server/TestCertificateServer.kt

```diff
@@ -43,63 +44,63 @@ class TestCertificateServer @Inject constructor(
     private val api: TestCertificateApiV1
         get() = dccApi.get()
 
-    @Throws(TestCertificateServerException::class)
+    @Throws(TestCertificateException::class)
     suspend fun registerPublicKeyForTest(
         testRegistrationToken: RegistrationToken,
         publicKey: RSAKey.Public,
     ): Unit = withContext(dispatcherProvider.IO) {
         Timber.tag(TAG).v("registerPublicKeyForTest(token=%s, key=%s)", testRegistrationToken, publicKey)
         if (!isInternetAvailable()) {
-            throw TestCertificateServerException(PKR_NO_NETWORK)
+            throw TestCertificateException(PKR_NO_NETWORK)
         }
         try {
-            val response = api.sendPublicKey(
+            api.sendPublicKey(
                 requestBody = TestCertificateApiV1.PublicKeyUploadRequest(
                     registrationToken = testRegistrationToken,
                     publicKey = publicKey.base64
                 )
             )
-            when (response.code()) {
-                400 -> throw TestCertificateServerException(PKR_400)
-                403 -> throw TestCertificateServerException(PKR_403)
-                404 -> throw TestCertificateServerException(PKR_404)
-                409 -> throw TestCertificateServerException(PKR_409)
-                500 -> throw TestCertificateServerException(PKR_500)
-            }
         } catch (e: Exception) {
             Timber.tag(TAG).w(e, "registerPublicKeyForTest failed")
             throw when (e) {
-                is TestCertificateServerException -> e
-                else -> TestCertificateServerException(PKR_FAILED, e)
+                is CwaWebException -> when (e.statusCode) {
+                    400 -> TestCertificateException(PKR_400, e)
```

Comment:
⬇️  imho it would be a little bit nice to use a networkInterceptor. Benefit would be to see the original response without okhttp interfering..
But optional. Will work this way too of course

Reply:
I contemplated this, but decided against it as it would make an already complex PR more complex: Interceptor vs just try+catch and the interceptor would "hide" within the module that creates the retrofit API, outside of the server class.

Overall I actually prefer the NetworkIntercepter approach, but we also already have a NetworkInterceptor for our `CwaWebExceptions`, introducing another one without more significant refactoring feels like there could be pitfalls not something I'd want to tackle now with the time pressure for 2.5.x.

My suggestion would be to keep this, and then refactor the whole handling independently. We use this kind of error mapping in other places too, that could then also get an interceptor, or we come up with a more elaborate pattern that can be reused.

@harambasicluka JIRA ticket for the refactoring epic? "Refactor HTTP error handling & mapping"?

Reply:
agreed.